### ISSUE TITLE
[AMDGPU] GFX11 disassembler doesn't support v_mov_b16 and v_swap_b16

### ISSUE THREAD

**Issue by DadSchoorse** [llvm/llvm-project#61622]
LLVM version: 3a3ad9fe1811ebd7ff865b77f471d155d9a05ec8

Example input: `7f043902`, which should be `v_mov_b16 v2, v2 opsel:[0, 0, 0, 1]` or `v_mov_b16 v2.h, v2.l`

**Comment by gandhi56** [comment]
`v_mov_b16 v2, v2 opsel:[0, 0, 0, 1]` does not appear to be supported by the assembler. llvm-mc reports them as invalid instructions:

```
/Users/gouenji56/workspace/llvm-project/llvm/test/MC/AMDGPU/gfx11_asm_vop1.s:2497:1: error: invalid instruction
v_swap_b16 v2, v2 opsel:[0, 0, 0, 1]
```

**Comment by DadSchoorse** [comment]
Yes, the assembler doesn't support `v_mov_b16` or `v_swap_b16` either as far as I can tell. I reported this as an issue with the disassembler, because that's the component I'm actually trying to use.

**Comment by DadSchoorse** [comment]
Any updates on this? It's pretty disappointing that LLVM still does not support such basic ALU opcodes. Failing to disassemble these instructions makes debugging RADV harder, and in some cases it even causes crashes in RadeonGPUProfiler. GFX11 isn't even that new anymore.

**Comment by gandhi56** [comment]
No updates, I can take a look sometime this week.

**Comment by jayfoad** [comment]
+ @Sisyph @kosarev

**Comment by kosarev** [comment]
Upstreaming the support for 16-bit instructions is still a work in progress; there are many 16-bit instructions that yet to be added, including `v_swap_b16`. `v_mov_b16` is supported and accepts `.l` and `.h` registers, so `v_mov_b16 v2.h, v2.l`.

Is there anything special about these two instructions? Adding `v_swap_b16` at the MC level should be trivial, so if this alone would resolve some practical problem on your end, I should be able to take a look.

**Comment by DadSchoorse** [comment]
> Upstreaming the support for 16-bit instructions is still a work in progress; there are many 16-bit instructions that yet to be added, including v_swap_b16.  Is there anything special about these two instructions? 

The difference is that the other opcodes are at least supported in some form, even if the disassembler will print warnings for invalid operands or won't print opsel. `v_swap_b16` is missing completely, which causes errors.

The issue with `v_mov_b16` is that it's not supported as fake16 as far as I understand, unlike all other 16bit true VOP1 which use the `VOP1Inst_t16` helper:
https://github.com/llvm/llvm-project/blob/eacefba9aa3d1a5181d3d49823df24aca0d2b344/llvm/lib/Target/AMDGPU/VOP1Instructions.td#L724

>  Adding v_swap_b16 at the MC level should be trivial, so if this alone would resolve some practical problem on your end, I should be able to take a look.

Yes, that would be enough for my use cases. At least I think so, MC level means assembler/disassembler, right?

**Comment by kosarev** [comment]
> MC level means assembler/disassembler, right?

Yes.

**Comment by kosarev** [comment]
@broxigarchen Guo, do you plan to upstream the contents of `gfx11_asm_t16.s` anytime soon? (I guess it should probably be part of some other test files.)

**Comment by broxigarchen** [comment]
Hi Ivan. That seems to be a file I missed. I will try to upstream it soon

**Comment by kosarev** [comment]
@broxigarchen Looks this can be marked resolved now?

**Comment by kosarev** [comment]
Re-assigned to Guo, to reflect the credit right.

**Comment by DadSchoorse** [comment]
This was implemented at some point, I think.
